### ISSUE TITLE
Pivot scheduling from polling to event-driven.

### ISSUE THREAD

**Issue by frankmcsherry** [frankmcsherry/timely-dataflow#117]
At the moment timely workers effectively poll the operators they manage, scheduling each operator once for each call to `worker.step()`. This works as long as you don't mind the wasted effort, but it is somewhat gross, can tie up time in larger dataflows, and recently has gotten in the way of logging as well (all of the pointless activations gum up the logs).

Instead, we could switch to an event-driven approach, where operators think a bit harder about whether they should be run at all. For example, operators with no incoming messages, no progress changes, and no work performed in their last invocation are probably up for being parked until something changes. Entire subgraphs can similarly be parked, so that we only drill down into parts of the dataflow graph that are currently active, or that need to be woken up for some reason.

I'm not 100% sure what those reasons should be at the moment, but my current thinking is:

Operators are executed when (i) there are inbound messages or progress changes, (ii) the operator has recently performed work (in its last invocation), or (iii) when the operator indicates that despite the absence of these it has some work it would like to do. In the absence of new stimulus, ongoing work, or explicit requests, the operator may be parked and re-executed only when they would receive additional stimulus.

1. At the moment workers are not explicitly told about incoming messages, from the communication threads, which is part of what would signal "new stimulus" (either in the form of data messages or progress messages). We could totally introduce this. My current thinking is that each worker could have a "todo" queue of operator identifiers, which can be written to by other threads. This would allow the communication thread to tickle operators without requiring them to upcall through owned data, and allow external threads to tickle operators when e.g. callbacks complete (e.g., async io to the filesystem or network). We log (and could surely provide to workers) the map from operator identifier to its "address" in the dataflow (a sequence of integer identifiers indicating its root-to-leaf path), from which the worker can determine which operators it should explore.

2. At the moment operators do not have a way to indicate that they are comfortable being de-scheduled or not. The `pull_internal_progress` call returns a boolean, which indicates whether the operator has any non-obvious work, i.e. should be scheduled again despite the absence of inbound frontier and held capabilities. This could be enriched to be an `Option<u64>` indicating a number of nanoseconds for which the operator is comfortable being descheduled; `Some(0)` would indicate "please reschedule asap", whereas `None` would indicate "no requirement to reschedule".

There is probably a bit of additional work ensuring that the `Subgraph` scheduling logic is comfortable with operators that may run varying numbers of times. Historically, the implementation has reset various buffers under the assumption that each operator has had a chance to react appropriately, and if nothing else some `assert!`s will probably need to be rethought. 

This doesn't seem like something that is going to land in a big hurry, but moving towards it in steps seems like a good idea.

**Comment by utaal** [comment]
A few months ago I prototyped a change that would let timely worker threads sleep if none of the three conditions were true:

>  (i) there are inbound messages or progress changes, (ii) the operator has recently performed work (in its last invocation), or (iii) when the operator indicates that despite the absence of these it has some work it would like to do.

My implementation was based on adding an `Activity` enum that was meant to encode (iii), similarly to your proposed `Option<u64>`:

```diff
+pub enum Activity {
+    /// No need to schedule again
+    Done,
+    /// Schedule again immediately
+    Internal,
+}
+
+impl Activity {
+    /// Combine this and other activity reports.
+    pub fn or(&self, other: Activity) -> Activity {
+        match (self, other) {
+            (&Activity::Internal, _) => Activity::Internal,
+            (_, Activity::Internal) => Activity::Internal,
+            (&Activity::Done, Activity::Done) => Activity::Done,
+        }
+    }
+}
```

which could be extended with a request to be re-scheduled at some point in the future.

The effect on operator logic looked something like this:

```diff
diff --git a/src/dataflow/operators/unary.rs b/src/dataflow/operators/unary.rs
index eb3ff94f..12f4547b 100644
--- a/src/dataflow/operators/unary.rs
+++ b/src/dataflow/operators/unary.rs
@@ -8,7 +8,7 @@ use progress::nested::subgraph::{Source, Target};

 use progress::count_map::CountMap;
 use dataflow::operators::Notificator;
-use progress::{Timestamp, Operate, Antichain};
+use progress::{Timestamp, Operate, Activity, Antichain};
 use dataflow::channels::pushers::Tee;
 use dataflow::channels::pushers::Counter as PushCounter;
 use dataflow::channels::pushers::buffer::Buffer as PushBuffer;
@@ -209,7 +209,7 @@ where T: Timestamp,

     fn pull_internal_progress(&mut self, consumed: &mut [CountMap<T>],
                                          internal: &mut [CountMap<T>],
-                                         produced: &mut [CountMap<T>]) -> bool
+                                         produced: &mut [CountMap<T>]) -> (bool, Activity)
     {
         {
             let mut input_handle = new_input_handle(&mut self.input, self.internal_changes.clone());
@@ -224,7 +224,7 @@ where T: Timestamp,
         self.output.inner().pull_progress(&mut produced[0]);
         self.internal_changes.borrow_mut().drain_into(&mut internal[0]);

-        false   // no unannounced internal work
+        (false, Activity::Done)   // no unannounced internal work
     }

     fn name(&self) -> String { self.name.clone() }
```

The rest of the diff (including some of the logic to summarise the `Activity` results for subgraphs) is here: https://github.com/frankmcsherry/timely-dataflow/compare/master...utaal:sleepy_timely

> We could totally introduce this. My current thinking is that each worker could have a "todo" queue of operator identifiers, which can be written to by other threads.

At the time, I went for the smallest change that I could make work, which seemed to be waking up all worker threads every time a message showed up on a communication thread (i.e. a communication thread was scheduled by the kernel). The "todo" queue approach sounds a lot cleaner, and may enable experimenting with different scheduling mechanisms. Additionally, it sounds like it could be more easily re-wired to work on top of non-blocking io.

**Comment by frankmcsherry** [comment]
Another condition that came up in conversation was "an operator that holds no capabilities and has no inbound messages", which might be grounds for parking. In a dataflow graph with good connectivity, any work at all may result in progress updates for everyone, but arguably these progress updates are not themselves a good reason to invoke an operator that does not actually have the ability to perform any dataflow work. At least, not without explicitly requesting such.

There are several features of an operator, and nailing down which combinations mean "needs to demand rescheduling" seems smart:

1. An operator may have input messages in queue.
2. An operator may hold capabilities.
3. An operator may have progress updates in queue.
4. An operator may have recently performed work.
5. An operator may have requested to be scheduled.

In conversation with @utaal, I think we came to the position that (4.) is interesting, but this could be policy pushed down to the operator implementation (e.g. `builder_rc`) where when an operator performs work it then requests rescheduling. This would allow other operators to not do so, if that makes more sense.

Looking at these, I would say that if (1.) holds the operator should be scheduled (otherwise, we have an operator that refuses to drain its input; could happen), if (2.) holds and (3.) happens the operator should be scheduled, as we have no information about the condition that might cause the operator to discharge a capability, and if (5.) happens just reschedule the operator.

**Comment by frankmcsherry** [comment]
Pondering this a bit more, and with a bit of experience trying to rewrite things, we may want a more rich vocabulary for describing scheduling interest. For example, the ubiquitous `probe` operator neither holds capabilities nor really receives messages, but it very much wants to be updated whenever its progress frontier changes. The ubiquitous `map` operator also holds no capabilities, but has no particular interest in being run when its progress frontier changes. These two operators look more-or-less identical to timely at the moment, and we should perhaps find a way to fix that. 

One minor distinction is that `map` has its `notify` bit set to false, which was previously a cute hack to optimize some progress traffic, but which might be an important signal.

It seems like there might be a more general vocabulary for when an operator is interested in being notified. One could imagine operators indicating that they (i) don't care about progress, (ii) always care about progress, (iii) care about progress when they have capabilities, or something more sophisticated. Derek was a fan of the idea that operators might advertise which frontiers they were interesting in, which might flow backwards through the graph, but something like `probe` would then probably taint the whole graph.

Anyhow, not yet clear how to avoid re-running every operator when an input advances having introduced no new records.

**Comment by qiuxiafei** [comment]
> if (2.) holds and (3.) happens the operator should be scheduled...

There's an exception in which operator should be scheduled only requires (2.) holds, it's the `source` operator. A `source` should be scheduled repeatedly until it holds no capability at all.

**Comment by frankmcsherry** [comment]
@qiuxiafei This is a good point, and something we've bumped in to a few times. At least, we've had a few iterations of "hrm, maybe these rules work?".

I think the current plan (written on a whiteboard only) is that the `dataplane_mockup` branch puts us on the path toward a world where each worker sees incoming messages explicitly, and can assemble a list of "operators that should run". More generally, there could be other ways to add operators to this list.

The input operators are good test cases, as I *think* what we want from them is the ability to tell us when they should run. That is, if the user has added no data nor advanced the timestamp, they don't need to be run. They only need to be run in response to the user actually doing something. This sounds minor, but there is a big binary difference between "nothing in the worker needs to run" and "one operator needs to be run repeatedly", in that it could let the worker thread sleep until an event arrives that would require an operator (or after a user-supplied timeout to let user code have a chance to run again).

But generally, the hope is that with a bit more work the dataplane branch can land and we can make informed scheduling (rather than just polling) be a higher priority. Until it lands, the worker doesn't have much insight into which operator don't need to run (as the comms thread can asynchronously populate a buffer).

**Comment by qiuxiafei** [comment]
Thanks, @frankmcsherry !
I think the problem has been cut into to parts:
1. How operators express their schedule interests.
2. How to schedule.

For (2.), maybe we can have a look at some actor system ([actix](https://actix.rs/) in rust for example). Actor systems have solved the 'stimulus' things, since message-driven, progress-driven and timer could be implemented on them directly. And even 'scheduling repeatedly' could be solved by sending a message to the actor its self. Just FYI.

And I'll check out the `dataplane_mockup` branch and go back later.

**Comment by zgramana** [comment]
There are some notes on approaches to managing `demand` (or not) in incremental computation systems in section 7 of this paper and which might be of some relevance to this discussion: 

 > [ADAPTON: Composable, Demand- Driven Incremental Computation](http://matthewhammer.org/adapton/adapton-pldi2014.pdf)

**Comment by frankmcsherry** [comment]
Closed with #206.
